### DISCUSSION TITLE
[V5] Asynchronously Load Env Vars Before Application Boots

### DISCUSSION THREAD

**Discussion by fareeda0** [adonisjs/core#2222] [Help] [answered]
So was playing around with setting environment variables (API keys etc.) before the application is booted. Managed to get this to work using a provider, but it's not 100%. Two main drawbacks that I can see are:

- `env.ts` validation doesn't work, as it will complain the variables are missing
- Any env vars used by `Config` provider will be `undefined`

Wondering if there any equivalent of Ignitor hooks that can be used asynchronously load and set these variables before the server and `Env` provider itself are loaded?

**Comment by thetutlage** [comment, marked as answer]
Can't comment much without looking at any code

**Reply by fareeda0** [reply]
Thanks for the reply, sure!

The provider looks more or less like this.


```ts
import { ApplicationContract } from '@ioc:Adonis/Core/Application'

export default class AppProvider {
  constructor (protected app: ApplicationContract) {
  }

  public register () {
    // Register your own bindings
  }

  public async boot () {
    // IoC container is ready
    const testVar: string = await new Promise((resolve) => setTimeout(() => resolve('my-secret-var'), 200))
    process.env['TEST_VAR'] = testVar
  }

  public async ready () {
    // App is ready
  }

  public async shutdown () {
    // Cleanup, since app is going down
  }
}
```

This sort of approach works so long as you don't have your env vars defined in the `.env.ts` file as you will get a `E_MISSING_ENV_VALUE` exception.

The other drawback of this is that any env vars used by the `Config` provider also come back as `undefined`.

I suspect the reason for this - which makes sense, is that the `Env` and `Config` providers are booted before this custom provider.

Are there any other lower-level hooks that I could use to set environment variables before the server is booted? 😄

**Reply by thetutlage** [reply]
I will not suggest writing this code. Here's why

The environment variables are the one that gets injected from the environment and must not be hardecoded within the code. Otherwise there is no value in use environment variables at all.

Feel free to reply, if you think I am missing something. But at first glance, environment variables should be defined inside code

**Reply by fareeda0** [reply]
I agree with this approach as well.

The only use case I had for this was for a few really sensitive environment variables that I didn’t want running around in plain text in my cloud provider UI.

Instead I opted to use a secret manager from my cloud provider and pull the environment variables on boot using a client library.

**Reply by thetutlage** [reply]
Okay. Makes sense and can that secret manager pull the environment variables on your server directly.

I understand the use case now, but the API is not async as of now and making it async requires some careful changes. So if they have an alternative loader for server, then you can use that too
